nowcasting_dataset/dataset/split/week.py

```diff
@@ -0,0 +1,108 @@
+from typing import Union, List, Tuple
+
+import random
+import numpy as np
+import pandas as pd
+
+
+def split_week(
+    datetimes: Union[List[pd.Timestamp], pd.DatetimeIndex],
+    train_test_validation_split: Tuple[int] = (3, 1, 1),
+) -> (List[pd.Timestamp], List[pd.Timestamp], List[pd.Timestamp]):
+    """
+    Split the data by the week.
+
+    If the split is (3,1,1) then
+    train data will have all weeks of the year that are module 5 remainder 0,1 or 2
+    validation data will have all weeks of the year that are module 5 remainder 3
+    test data will have all weeks of the year that are module 5 remainder 4
+
+    Args:
+        datetimes: list of datetimes
+        train_test_validation_split: how the split is made
+
+    Returns: train, validation and test datetimes
+
+    """
+
+    total_weights = sum(train_test_validation_split)
+    cum_weights = np.cumsum(train_test_validation_split)
+
+    # make which week indexes go i.e if the split is [3,1,1] then the
+    # - train_indexes = [0,1,2]
+    # - validation_indexes = [3]
+    # - test_indexes = [4]
+    train_indexes = [i for i in range(total_weights) if i < cum_weights[0]]
+    validation_indexes = [
+        i for i in range(total_weights) if (i >= cum_weights[0]) & (i < cum_weights[1])
+    ]
+    test_indexes = [i for i in range(total_weights) if (i >= cum_weights[1]) & (i < cum_weights[2])]
+
+    # select the datetimes
+    train = [
+        datetime for datetime in datetimes if datetime.weekofyear % total_weights in train_indexes
```

Comment:
Similarly to the comments on `day.py`, I'd recommend filtering on _all_ the weeks in the dataset.  You can get all the unique weeks in the dataset using `unique_weeks = np.unique(datetime.to_period('W')`.

And, if you use `datetime.to_period(freq)` then that perhaps opens up the possibility of reducing code duplication by using the same function for splitting by weeks or days or months or years, by passing in `freq='D'` or `W` or `M` or `Y` :)

Reply:
Yea, ill join it all together. Good idea with the 'freq'